src/analyz/ty_json.rs

```diff
@@ -1135,6 +936,408 @@ impl<'tcx> ToJson<'tcx> for ty::Const<'tcx> {
+                    "elements": elt_values
+                }))
+            },
+            _ => ()
+        }
+    }
+
+    let (prov, _offset) = d.ptr.into_parts();
+    let alloc = tcx.try_get_global_alloc(prov?)?;
+    match alloc {
+        interpret::GlobalAlloc::Static(def_id) =>
+            return Some(json!({
+                "kind": "static_ref",
+                "def_id": def_id.to_json(mir),
+            })),
+        interpret::GlobalAlloc::Memory(ca) => {
+            let ty = op_ty.layout.ty;
+            let aid = match mir.allocs.get(ca, ty) {
+                Some(alloc_id) => alloc_id.to_owned(),
+                None => {
+                    let ma = tcx.create_memory_alloc(ca);
+                    let rlayout = tcx.layout_of(ty::ParamEnv::reveal_all().and(rty)).unwrap();
+                    let ptr = interpret::Pointer::new(Some(ma), Size::ZERO);
+
+                    let mpty = interpret::MPlaceTy::from_aligned_ptr_with_meta(ptr, rlayout, d.meta);
+                    let rendered = try_render_opty(mir, icx, &mpty.into());
+
+                    let static_ref = json!({
+                        "kind": "constant",
+                        "mutable": false,
+                        "ty": rty.to_json(mir),
+                        "rendered": rendered,
+                    });
+
+                    mir.allocs.insert(tcx, ca, ty, static_ref)
+                }
+            };
+
+            return Some(json!({
+                "kind": "static_ref",
+                "def_id": aid,
+            }));
+        }
+        _ => return None
+    }
+}
+
+// A copied version of MPlaceTy::len, which (sadly) isn't exported. See
+// https://github.com/rust-lang/rust/blob/56ee85274e5a3a4dda92f3bf73d1664c74ff9c15/compiler/rustc_const_eval/src/interpret/place.rs#L227C5-L243
+#[inline]
+pub fn mplace_ty_len<'tcx, Tag: Provenance>(mplace_ty: &MPlaceTy<'tcx, Tag>, cx: &impl HasDataLayout) -> InterpResult<'tcx, u64> {
+    if mplace_ty.layout.is_unsized() {
+        // We need to consult `meta` metadata
+        match mplace_ty.layout.ty.kind() {
+            ty::Slice(..) | ty::Str => mplace_ty.meta.unwrap_meta().to_machine_usize(cx),
+            _ => bug!("len not supported on unsized type {:?}", mplace_ty.layout.ty),
+        }
+    } else {
+        // Go through the layout.  There are lots of types that support a length,
+        // e.g., SIMD types. (But not all repr(simd) types even have FieldsShape::Array!)
+        match mplace_ty.layout.fields {
+            FieldsShape::Array { count, .. } => Ok(count),
+            _ => bug!("len not supported on sized type {:?}", mplace_ty.layout.ty),
+        }
+    }
+}
+
+// TODO: move
+pub fn mk_interp<'mir, 'tcx>(tcx: TyCtxt<'tcx>) -> interpret::InterpCx<'mir, 'tcx, RenderConstMachine<'mir, 'tcx>> {
+    interpret::InterpCx::new(
+        tcx,
+        DUMMY_SP,
+        ty::ParamEnv::reveal_all(),
+        RenderConstMachine::new(),
+    )
+}
```

Comment:
Seems this function is only called in one place, so the logic could be inlined there.